linera-execution/src/lib.rs

```diff
@@ -1027,6 +1027,37 @@ pub struct BlobState {
     pub epoch: Epoch,
 }
 
+/// Responses to oracle queries that are being recorded or replayed.
+#[derive(Debug)]
+pub enum OracleResponses {
+    /// When executing a block _proposal_, oracles can be used and their responses are recorded.
+    Record(Vec<OracleResponse>),
+    /// When re-executing a validated or confirmed block, recorded responses are used.
+    Replay(vec::IntoIter<OracleResponse>),
+    /// In service queries, oracle responses are not recorded.
+    Forget,
+}
+
+impl OracleResponses {
+    /// Returns an empty `Record` if the argument is `None`, and `Replay` if it is `Some`.
+    pub fn from_record(oracle_record: Option<OracleRecord>) -> Self {
+        if let Some(responses) = oracle_record {
+            OracleResponses::Replay(responses.responses.into_iter())
+        } else {
+            OracleResponses::Record(Vec::new())
+        }
+    }
+
+    /// Returns the oracle record if `self` is `Record` and an empty record otherwise.
+    pub fn into_record(self) -> OracleRecord {
```

Comment:
I feel that the default value should be explicit, maybe `tape.[try_into_]record().or_default()`